test/E2E.test.ts

```diff
@@ -0,0 +1,419 @@
+  OFFCHAIN_CONFIG: "0x",
+};
+
+describe("DssCronKeeper KeeperRegistry", function () {
+  let keeper: DssCronKeeper;
+  let sequencer: Sequencer;
+  let registry: KeeperRegistry20;
+  let registrar: KeeperRegistrar20;
+  let linkToken: LinkTokenMock;
+  let upkeepId: string;
+  let vestId: BigNumber;
+  let owner: SignerWithAddress;
+  let registrySigners: Wallet[];
+  let dssVest: DssVestMintable;
+  let topUp: DssVestTopUp;
+  let paymentToken: ERC20PresetMinterPauser;
+
+  if (
+    !STAGING_LINK_TOKEN ||
+    !STAGING_LINK_USD_PRICE_FEED ||
+    !STAGING_PAYMENT_USD_PRICE_FEED ||
+    !STAGING_SWAP_ROUTER ||
+    !VOW ||
+    !KEEPER_REGISTRY_LOGIC ||
+    !NONFUNGIBLE_POSITION_MANAGER
+  ) {
+    throw new Error("Missing required env variables!");
+  }
+
+  before(async function () {
+    [owner] = await ethers.getSigners();
+
+    linkToken = await ethers.getContractAt("LinkTokenMock", STAGING_LINK_TOKEN);
+
+    paymentToken = await setupPaymentToken(owner);
+    registrySigners = await getRegistrySigners();
+  });
+
+  beforeEach(async function () {
+    sequencer = await setupSequencer();
+
+    await setupSampleJob(sequencer);
+
+    keeper = await setupKeeper(sequencer);
+
+    ({ registry, registrar } = await setupChainlinkContracts(owner, linkToken));
+
+    upkeepId = await registerUpkeep(
+      keeper.address,
+      linkToken.address,
+      registrar.address,
+      owner.address,
+      UpkeepParams.ADMIN_EMAIL,
+      UpkeepParams.NAME,
+      UpkeepParams.GAS_LIMIT,
+      UpkeepParams.OFFCHAIN_CONFIG,
+      UpkeepParams.INITIAL_FUNDING,
+      UpkeepParams.CHECK_DATA
+    );
+
+    dssVest = await setupDssVest(paymentToken);
+
+    const daiJoinMock = await setupDaiJoin(paymentToken);
+
+    topUp = await setupDssVestTopup(
+      dssVest,
+      daiJoinMock,
+      paymentToken,
+      registry,
+      linkToken,
+      upkeepId,
+      keeper,
+      VOW,
+      STAGING_SWAP_ROUTER,
+      STAGING_PAYMENT_USD_PRICE_FEED,
+      STAGING_LINK_USD_PRICE_FEED
+    );
+
+    vestId = await setupVest(dssVest, topUp, owner, vestId);
+
+    // fund registrySigner1 with some ETH so it can execute performUpkeep
+    await owner.sendTransaction({
+      to: registrySigners[0].address,
+      value: ethers.utils.parseEther("1.0"),
+    });
+  });
+
+  describe("upkeep", function () {
+    it("should have created new upkeep", async function () {
+      expect(upkeepId).to.not.eq(undefined);
+    });
+  });
+
+  describe("Execute jobs", function () {
+    it("should reduce upkeep balance after job execution", async function () {
+      const { balance: preBalance } = await registry.getUpkeep(upkeepId);
+      await keeperRegistryPerformUpkeep(
+        registry,
+        registrySigners,
+        upkeepId,
+        KeeperRegistryParams.F
+      );
+      const { balance: postBalance } = await registry.getUpkeep(upkeepId);
+      expect(preBalance).to.be.gt(postBalance);
+    });
+  });
+
+  describe("Refund upkeep", function () {
```

Comment:
It would be nice to have checks for the uniswap integration, like is the expected amount of tokens were received.